src/main/java/frc/robot/subsystems/drive/SwerveModule.java

```diff
@@ -0,0 +1,220 @@
+    drivingEncoderTmp.setVelocityConversionFactor(
+        Constants.SwerveModule.DRIVING_ENCODER_VELOCITY_FACTOR_METERS_PER_SECOND);
+
+    drivingSparkMax.setIdleMode(Constants.SwerveModule.DRIVING_MOTOR_IDLE_MODE);
+    drivingSparkMax.setSmartCurrentLimit(
+        Constants.SwerveModule.DRIVING_MOTOR_CURRENT_LIMIT_AMPS);
+  }
+
+  /**
+   * Burns the current settings to sparks so they keep current settings on reboot. Should be done
+   * after all settings are set.
+   */
+  public void burnFlashSparks() {
+    Timer.delay(0.005);
+    drivingSparkMax.burnFlash();
+    Timer.delay(0.005);
+    turningSparkMax.burnFlash();
+  }
+
+  /**
+   * Returns the current state of the module.
+   *
+   * @return The current state of the module.
+   */
+  public SwerveModuleState getState() {
+    // Apply chassis angular offset to the encoder position to get the position
+    // relative to the chassis.
+    return new SwerveModuleState(
+        drivingEncoder.getVelocity(),
+        new Rotation2d(turningEncoder.getPosition() - chassisAngularOffsetRadians));
+  }
+
+  /**
+   * Returns the current position of the module.
+   *
+   * @return The current position of the module.
+   */
+  public SwerveModulePosition getPosition() {
+    // Apply chassis angular offset to the encoder position to get the position
+    // relative to the chassis.
+    return new SwerveModulePosition(
+        drivingEncoder.getPosition(),
+        new Rotation2d(turningEncoder.getPosition() - chassisAngularOffsetRadians));
+  }
+
+  /**
+   * Sets the desired state for the module.
+   *
+   * @param desiredState Desired state with speed and angle.
+   */
+  public void setDesiredState(SwerveModuleState desiredState) {
+    // Apply chassis angular offset to the desired state.
+    SwerveModuleState correctedDesiredState = new SwerveModuleState();
+    correctedDesiredState.speedMetersPerSecond = desiredState.speedMetersPerSecond;
+    correctedDesiredState.angle =
+        desiredState.angle.plus(Rotation2d.fromRadians(chassisAngularOffsetRadians));
+
+    // Optimize the reference state to avoid spinning further than 90 degrees.
+    SwerveModuleState optimizedDesiredState =
+        SwerveModuleState.optimize(
+            correctedDesiredState, new Rotation2d(turningEncoder.getPosition()));
+
+    // Setting global desiredState to be optimized for the shuffleboard
+    this.desiredState = optimizedDesiredState;
+
+    // Command driving and turning SPARKS MAX towards their respective setpoints.
+    drivingPIDController.setReference(
+        optimizedDesiredState.speedMetersPerSecond, CANSparkMax.ControlType.kVelocity);
+    turningPIDController.setReference(
+        optimizedDesiredState.angle.getRadians(), CANSparkMax.ControlType.kPosition);
+  }
+
+  /** Zeroes all the SwerveModule encoders. */
+  public void resetEncoders() {
+    drivingEncoder.setPosition(0);
+  }
+
+  public double getEncoderAbsPositionRad() {
+    return turningEncoder.getPosition();
```

Comment:
I could be wrong, should this be renamed to getTurningEncoderAbsPositionRad? Or does it not matter?